[sram_ctrl] Harden initialization counter

msfschaffner opened the pull request:
This duplicates the initialization counter and slightly reworks the escalation logic.

Counter mismatches now also trigger an alert, and local countermeasures such as key scrapping, nonce reversal and transaction blocking.

Signed-off-by: Michael Schaffner <msf@google.com>

udinator commented:
@msfschaffner could you please run the `lc_escalation` test just to confirm that nothing has broken there?

msfschaffner commented:
Thanks @udinator - there were indeed some issues that needed fixing:
- CSR prediction was not entirely correct (both on DV and Design side). I updated the CSR behavior such that both key valid and the init done bits are cleared upon escalation.
- The `req_mem_init` function needed to be updated, since the programming model for that has changed (need to spinwait on a different CSR bit).
- I ran into another RAW corner case, but this time this manifested itself through a different sequence (will note it on #7461). I modified the existing workaround #7461 to account for this.

After these fixes, some `lc_escalation` sequences now error out with a `Timeout waiting tl_access` error that I am not entirely sure how to fix. This timeout error occurs in case a subword write is accepted in the address phase, and escalation is asserted between the read and write operations of the read-modify-write operation, basically causing the transaction to get stuck (and the TB still thinks the transaction has entered the SRAM pipeline before escalation has taken effect).

udinator reviewed (approved):
LGTM, thanks for fixing up the DV side of things.
